clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go

```diff
@@ -23,239 +23,410 @@ package common
 	if err != nil {
 		return summaries, err
 	}
-	namespace, err := util.GetStringOrDefault(config.Params, "namespace", metav1.NamespaceAll)
-	if err != nil {
-		return summaries, err
+
+	switch action {
+	case "start":
+		w.kind, err = util.GetString(config.Params, "kind")
+		if err != nil {
+			return summaries, err
+		}
+		w.namespace, err = util.GetStringOrDefault(config.Params, "namespace", metav1.NamespaceAll)
+		if err != nil {
+			return summaries, err
+		}
+		w.labelSelector, err = util.GetStringOrDefault(config.Params, "labelSelector", "")
+		if err != nil {
+			return summaries, err
+		}
+		w.fieldSelector, err = util.GetStringOrDefault(config.Params, "fieldSelector", "")
+		if err != nil {
+			return summaries, err
+		}
+		w.operationTimeout, err = util.GetDurationOrDefault(config.Params, "operationTimeout", defaultOperationTimeout)
+		if err != nil {
+			return summaries, err
+		}
+		return summaries, w.start(config.ClientSet)
+	case "gather":
+		timeout, err := util.GetDurationOrDefault(config.Params, "timeout", defaultWaitForRCPodsTimeout)
+		if err != nil {
+			return summaries, err
+		}
+		return summaries, w.gather(config.ClientSet, timeout)
+	case "stop":
+		w.stop()
+		return summaries, nil
+	default:
+		return summaries, fmt.Errorf("unknown action %v", action)
 	}
-	labelSelector, err := util.GetStringOrDefault(config.Params, "labelSelector", "")
-	if err != nil {
-		return summaries, err
+}
+
+func (w *waitForControlledPodsRunningMeasurement) start(c clientset.Interface) error {
+	if w.informer != nil {
+		glog.Infof("%v: wait for controlled pods measurement already running", waitForControlledPodsRunningName)
+		return nil
 	}
-	fieldSelector, err := util.GetStringOrDefault(config.Params, "fieldSelector", "")
-	if err != nil {
-		return summaries, err
+	glog.Infof("%v: starting wait for controlled pods measurement...", waitForControlledPodsRunningName)
+	optionsModifier := func(options *metav1.ListOptions) {
+		options.FieldSelector = w.fieldSelector
+		options.LabelSelector = w.labelSelector
 	}
-	timeout, err := util.GetDurationOrDefault(config.Params, "timeout", defaultWaitForRCPodsTimeout)
+	listerWatcher := cache.NewFilteredListWatchFromClient(c.CoreV1().RESTClient(), w.kind+"s", w.namespace, optionsModifier)
+	obj, err := runtimeobjects.GetObjectForKind(w.kind)
 	if err != nil {
-		return summaries, err
+		return err
+	}
+	w.isRunning = true
+	w.stopCh = make(chan struct{})
+	w.informer = cache.NewSharedInformer(listerWatcher, obj, 0)
+	w.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			addF := func() {
+				runtimeObj, ok := obj.(runtime.Object)
+				if !ok {
+					return
+				}
+				if err := w.updateOpResourceVersion(runtimeObj); err != nil {
+					return
+				}
+				w.handleObject(c, runtimeObj)
+			}
+			w.queue.Add(&addF)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			updateF := func() {
+				oldRuntimeObj, ok := oldObj.(runtime.Object)
```

Comment:
Can you opaque this into function (and share that between add/update/delete).

```
func (w *...) handleUpdateObject(oldObj, newObj interface{}) {
   // what you have here (maybe you need to handle nil differently, but I think it will work)
}
```

And then here:

```
AddFunc: func(obj interface{}) {
  w.queue.Add(w.handleUpdateObject(nil, obj)
}
...
```

Reply:
> Can you opaque this into function (and share that between add/update/delete).

Done.

```
AddFunc: func(obj interface{}) {
  w.queue.Add(w.handleUpdateObject(nil, obj)
}
...
```
Not sure how to do this:
- Function needs to be called after retrieving it from queue, not before. I can change `handleObject` so it returns `func()` but I'm not sure if this will improve readability.
- Due to queue using map internally, function are not support. Although, there is no problem with function pointers.